test/e2e/failover_test.go

```diff
@@ -242,6 +242,234 @@ var _ = framework.SerialDescribe("failover testing", func() {
+					ResourceSelectors: []policyv1alpha1.ResourceSelector{
+						{
+							APIVersion: deployment.APIVersion,
+							Kind:       deployment.Kind,
+							Name:       deployment.Name,
+						},
+					},
+					Placement: policyv1alpha1.Placement{
+						ClusterAffinity: &policyv1alpha1.ClusterAffinity{
+							ClusterNames: framework.ClusterNames(),
+						},
+						SpreadConstraints: []policyv1alpha1.SpreadConstraint{
+							{
+								SpreadByField: policyv1alpha1.SpreadByFieldCluster,
+								MaxGroups:     maxGroups,
+								MinGroups:     minGroups,
+							},
+						},
+					},
+					PropagateDeps: true,
+					Failover: &policyv1alpha1.FailoverBehavior{
+						Application: &policyv1alpha1.ApplicationFailoverBehavior{
+							DecisionConditions: policyv1alpha1.DecisionConditions{
+								TolerationSeconds: pointer.Int32(60),
+							},
+							PurgeMode: policyv1alpha1.Never,
+						},
+					},
+				},
+			}
+		})
+
+		ginkgo.BeforeEach(func() {
+			framework.CreatePropagationPolicy(karmadaClient, policy)
+			framework.CreateDeployment(kubeClient, deployment)
+			ginkgo.DeferCleanup(func() {
+				framework.RemoveDeployment(kubeClient, deployment.Namespace, deployment.Name)
+				framework.RemovePropagationPolicy(karmadaClient, policy.Namespace, policy.Name)
+			})
+		})
+
+		ginkgo.It("application failover", func() {
+			disabledClusters := framework.ExtractTargetClustersFrom(controlPlaneClient, deployment)
+			ginkgo.By("create an error op", func() {
+				overridePolicy = testhelper.NewOverridePolicyByOverrideRules(policyNamespace, policyName, []policyv1alpha1.ResourceSelector{
+					{
+						APIVersion: deployment.APIVersion,
+						Kind:       deployment.Kind,
+						Name:       deployment.Name,
+					},
+				}, []policyv1alpha1.RuleWithCluster{
+					{
+						TargetCluster: &policyv1alpha1.ClusterAffinity{
+							ClusterNames: disabledClusters,
+						},
+						Overriders: policyv1alpha1.Overriders{
+							ImageOverrider: []policyv1alpha1.ImageOverrider{
+								{
+									Component: "Registry",
+									Operator:  "replace",
+									Value:     "fake",
+								},
+							},
+						},
+					},
+				})
+				framework.CreateOverridePolicy(karmadaClient, overridePolicy)
+			})
+
+			ginkgo.By("check whether the failed deployment is rescheduled to other available cluster", func() {
+				gomega.Eventually(func() int {
+					targetClusterNames := framework.ExtractTargetClustersFrom(controlPlaneClient, deployment)
+					for _, targetClusterName := range targetClusterNames {
+						// the target cluster should be overwritten to another available cluster
+						if !testhelper.IsExclude(targetClusterName, disabledClusters) {
+							return 0
+						}
+					}
+
+					return len(targetClusterNames)
+				}, pollTimeout, pollInterval).Should(gomega.Equal(minGroups))
+				framework.WaitDeploymentPresentOnClustersFitWith(disabledClusters, deploymentNamespace, deploymentName, func(deployment *appsv1.Deployment) bool { return true })
```

Comment:
We can use the `ginkgo.By` module to organize the test separately.

Reply:
Is it like:
```
ginkgo.By("xxxx", func(){
     framework.WaitDeploymentPresentOnClustersFitWith(disabledClusters, deploymentNamespace, deploymentName, func(deployment *appsv1.Deployment) bool {return true} )
})
```

Reply:
Yes, it seems better.